xarray/core/dataset.py

```diff
@@ -2748,6 +2748,64 @@ def roll(self, **shifts):
 
         return self._replace_vars_and_dims(variables)
 
+    def sortby(self, variables, ascending=True):
+        """
+        Sorts the dataset, either along specified dimensions,
+        or according to values of 1-D dataarrays that share dimension
+        with calling object.
+
+        Parameters
+        ----------
+        variables: (str, DataArray, or iterable of either)
+            Name of a 1D variable in coords/data_vars whose values are used to
+            to sort the dataset.
+        ascending: boolean, optional.
+            whether to sort by ascending order.
+
+        Returns
+        -------
+        sorted: Dataset
+            A new dataset where all the specified dims are sorted by dim
+            labels.
+        """
+        from .dataarray import DataArray
+        if isinstance(variables, (str, unicode, DataArray)):
+            vs = [variables]
+        elif not isinstance(variables, Iterable):
+            raise TypeError("variables must be either str or iterable."
+                            "  Got {}.".format(type(variables)))
+        else:  # an iterable
+            if not all([isinstance(v, (str, unicode, DataArray))
+                       for v in variables]):
+                raise TypeError("At least one of the variables is not "
+                                "str or DataArray.")
+            else:
+                vs = variables
+        dims = []
+        for d in vs:
+            if isinstance(d, (str or unicode)):
+                if d not in self:
+                    raise KeyError("Dimension {} does not exist "
+                                   "in object.".format(d))
+                key = d
+                val = self[d].argsort() if ascending\
+                    else self[d].argsort()[::-1]
+            else:  # d is a DataArray
+                if len(d.dims) > 1:
+                    raise ValueError("Input DataArray has more "
+                                     "than 1 dimension.")
+                else:
+                    key = d.dims[0]
+                    if key not in self:
+                        raise KeyError("Dimension {} does not exist "
+                                       "in object.".format(key))
+                    val = d.argsort() if ascending else d.argsort()[::-1]
+                    if len(val) != len(self[key]):
+                        raise ValueError("Input DataArray must have same "
+                                         "length as dimension it is sorting.")
+            dims.append((key, val))
+        return self.isel(**dict(dims))
```

Comment:
What happens if there are multiple keys provided along the same dimension? This is equivalent to sorting by multiple columns in a spreadsheet.

It's OK not to support it, but we should be sure to raise an informative error, because I'm sure somebody will try it. Note that when you put multiple items in `dict` with the same name Python *ignores* the first argument:
```
In [1]: dict([('x', 1), ('x', 2)])
Out[1]: {'x': 2}
```

If you *do* want to support it, probably the easiest way is with [`np.lexsort`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.lexsort.html)